src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/RuntimeAsyncRewriter.cs

```diff
@@ -112,4 +116,82 @@ public BoundExpression VisitExpression(BoundExpression node)
         // System.Runtime.CompilerServices.RuntimeHelpers.Await(awaitedExpression)
         return _factory.Call(receiver: null, awaitMethod, VisitExpression(node.Expression));
     }
+
+    private BoundExpression RewriteCustomAwaiterAwait(BoundAwaitExpression node)
+    {
+        // await expr
+        // becomes
+        // var _tmp = expr.GetAwaiter();
+        // if (!_tmp.IsCompleted)
+        //    UnsafeAwaitAwaiterFromRuntimeAsync(_tmp) OR AwaitAwaiterFromRuntimeAsync(_tmp);
+        // _tmp.GetResult();
+
+        // PROTOTYPE: await dynamic will need runtime checks, see AsyncMethodToStateMachine.GenerateAwaitOnCompletedDynamic
+
+        var expr = VisitExpression(node.Expression);
+
+        var awaitablePlaceholder = node.AwaitableInfo.AwaitableInstancePlaceholder;
+        if (awaitablePlaceholder is not null)
+        {
+            _placeholderMap.Add(awaitablePlaceholder, expr);
+        }
+
+        // expr.GetAwaiter()
+        var getAwaiter = VisitExpression(node.AwaitableInfo.GetAwaiter);
+        Debug.Assert(getAwaiter is not null);
+
+        if (awaitablePlaceholder is not null)
+        {
+            _placeholderMap.Remove(awaitablePlaceholder);
+        }
+
+        // var _tmp = expr.GetAwaiter();
+        var tmp = _factory.StoreToTemp(getAwaiter, out BoundAssignmentOperator store, kind: SynthesizedLocalKind.Awaiter);
+
+        // _tmp.IsCompleted
+        var isCompletedMethod = node.AwaitableInfo.IsCompleted!.GetMethod;
```

Comment:
>! [](http://example.com/codeflow?start=62&length=1)

Unless there's an earlier check in the method that implies this, an assertion is preferred
From our [guideline](https://github.com/dotnet/roslyn/blob/main/docs/contributing/Nullable%20Annotations.md): "DO use the null suppression operator in places where a method body has ensured a value cannot be null but the compiler is unable to track it"